File: app/src/main/java/com/sakurafish/pockettoushituryou/util/AlarmUtils.java
package com.sakurafish.pockettoushituryou.util;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;

import com.sakurafish.pockettoushituryou.receiver.NotificationReceiver;

import java.util.Calendar;

public class AlarmUtils {

    private static final String TAG = AlarmUtils.class.getSimpleName();

    public static void registerAlarm(@NonNull Context context) {
        // Wake Up every day at PM12:00 for app message
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (alarmManager == null) return;

        // every day at scheduled time
        Calendar calendar = Calendar.getInstance();
        // if it's after or equal 12 pm schedule for next day
        if (Calendar.getInstance().get(Calendar.HOUR_OF_DAY) >= 12) {
            calendar.add(Calendar.DAY_OF_YEAR, 1);
        }
        calendar.set(Calendar.HOUR_OF_DAY, 12);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);

        PendingIntent pendingIntent = createAlarmIntent(context);
        alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP,
                calendar.getTimeInMillis(), AlarmManager.INTERVAL_DAY, pendingIntent);
    }

    public static void unregisterAlarm(@NonNull Context context) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (alarmManager == null) return;

        PendingIntent pendingIntent = createAlarmIntent(context);
        pendingIntent.cancel();
        alarmManager.cancel(pendingIntent);
    }

    private static PendingIntent createAlarmIntent(@NonNull Context context) {
        Intent pendingIntent = NotificationReceiver.createIntent(context);
        return PendingIntent.getBroadcast(context, NotificationReceiver.NOTIFICATION_ID,
                pendingIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }
}

File: app/src/main/java/com/sakurafish/pockettoushituryou/shared/AlarmUtils.kt
package com.sakurafish.pockettoushituryou.shared

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import com.sakurafish.pockettoushituryou.receiver.NotificationReceiver
import java.util.*

object AlarmUtils {
    private val TAG = AlarmUtils::class.java.simpleName

    @JvmStatic
    fun registerAlarm(context: Context) {
        // Wake Up every day at PM12:00 for app message
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

        // every day at scheduled time
        val calendar = Calendar.getInstance()
        // if it's after or equal 12 pm schedule for next day
        if (Calendar.getInstance()[Calendar.HOUR_OF_DAY] >= 12) {
            calendar.add(Calendar.DAY_OF_YEAR, 1)
        }
        calendar[Calendar.HOUR_OF_DAY] = 12
        calendar[Calendar.MINUTE] = 0
        calendar[Calendar.SECOND] = 0
        val pendingIntent = createAlarmIntent(context)
        alarmManager.setInexactRepeating(
            AlarmManager.RTC_WAKEUP,
            calendar.timeInMillis, AlarmManager.INTERVAL_DAY, pendingIntent
        )
    }

    @JvmStatic
    fun unregisterAlarm(context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = createAlarmIntent(context)
        pendingIntent.cancel()
        alarmManager.cancel(pendingIntent)
    }

    private fun createAlarmIntent(context: Context): PendingIntent {
        val pendingIntent = NotificationReceiver.createIntent(context)
        val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE

        return PendingIntent.getBroadcast(
            context, NotificationReceiver.NOTIFICATION_ID,
            pendingIntent, flags
        )
    }
}